cpp/src/arrow/array/array_nested.cc

```diff
@@ -847,14 +847,13 @@ Result<std::shared_ptr<Array>> MapArray::FromArraysInternal(
   const auto& typed_offsets = checked_cast<const OffsetArrayType&>(*offsets);
 
   BufferVector buffers;
-  int64_t null_count;
-  if (null_bitmap != nullptr) {
-    buffers = BufferVector({std::move(null_bitmap), typed_offsets.values()});
-    null_count = null_bitmap->size();
-  } else {
-    buffers = BufferVector({null_bitmap, typed_offsets.values()});
-    null_count = 0;
+  buffers.resize(2);
+  int64_t null_count = 0;
+  if (null_bitmap) {
+    buffers[0] = std::move(null_bitmap);
+    null_count = internal::CountSetBits(null_bitmap->data(), 0, offsets->length());
```

Comment:
As mentioned in the other PR, you want to count the "unset" bits. I think typically that is done by doing `null_count = length - CountSetBits(..)`. But, maybe we can also just keep this defined at `-1` (as it was before) to postpone the actual count.

Reply:
In any case it would be good to expand the test case to actually test the null_count of the result to ensure this is covered and we would have caught the initial mistake

Reply:
(Python tests caught the error I have made in `null_count` calculation - I forgot to run it locally 😊 )

Reply:
Also: the expected MapArray constructed in the C++ tests checks for `null_count`.  I have mistakenly specified 1 instead of 3 as the `offsets_is_valid3`is defined as `{true, false, true, true}` and I have misunderstood the missing flag). Will add separate check for `null_count` nevertheless.